### PR TITLE
Add a12n-server authentication #13

### LINKED ISSUE #13: Add a12n-server authentication

### PR DESCRIPTION
**Pull request by syedfkabir**
Resolves #13 

## Changes

- added a12n server authentication 
- added a semicolon to a statement that needed one
- added several new dependencies to enable OAuth2
- updated the `README.md` to include a redirect that auto completes the a12n principal app setup

## Screenshots
Screen Shot 2022-09-30 at 12 41 55 PM

## Note 
This can't be merged without the corresponding front end ticket being updated as well.

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -46,6 +60,27 @@ app.use(validator({
   schemaPath: path.join(__dirname, '../node_modules/@badgateway/tt-types/schema')
 }));
 
+// a12n setup
+const client = new OAuth2Client({
+  server: 'http://localhost:8531',
```
Comment (evert) [inline src/app.ts:65]: This shouldn't be hardcoded

File: src/app.ts [hunk 2]
```diff
@@ -46,6 +60,27 @@ app.use(validator({
...
+  clientId: 'tt-api',
```
Comment (evert) [inline src/app.ts:66]: I thought this was also in the .env file?

File: src/app.ts [hunk 3]
```diff
@@ -46,6 +60,27 @@ app.use(validator({
...
+  clientSecret: process.env.OAUTH2_CLIENT_SECRET,
+  tokenEndpoint: '/token',
+  introspectionEndpoint: '/introspect',
```
Comment (evert) [inline src/app.ts:69]: These 2 don't need to be specified, they should get auto-detected if the `server` is specified correctly.

File: src/app.ts [hunk 4]
```diff
@@ -46,6 +60,27 @@ app.use(validator({
...
+});
+
+app.use(browserToBearer({client}));
+
+app.use(oauth2({
+  publicPrefixes: [
+    '/health-check',
+    '/login',
+    '/register'
```
Comment (evert) [inline src/app.ts:78]: Where do these come from?

**Review by evert — CHANGES_REQUESTED** [review]
This PR ideally should also update the readme with the automated setup link so people have an easy time configuring their a12n-server

File: src/app.ts (hunk 4 above)
Comment (syedfkabir) [reply src/app.ts:78]: not sure, I was doing a lot of copying and pasta'ing from `media-api`

i see there were a lot of things I missed! another commit coming soon!

File: src/app.ts [hunk 5]
```diff
@@ -46,6 +60,23 @@ app.use(validator({
...
+  server: process.env.AUTH_API_URI,
+  clientId: process.env.OAUTH2_CLIENT_ID || 'tt-api',
+  clientSecret: process.env.OAUTH2_CLIENT_SECRET,
+});
+
+app.use(browserToBearer({client}));
+
+app.use(oauth2({
+  publicPrefixes: [
+    '/health-check',
```
Comment (evert) [inline src/app.ts:74]: Our standard endpoint for this is `/health`, but looks good otherwise!

```suggestion
    '/health',
```

**Review by evert — APPROVED** [review]

**Review by pschwyter — APPROVED** [review]

**Pull request merged**